Add API versioning

OleksandrOrlov opened the pull request:
The PR adds version number in resource URIs. This will allow clients to distinguish between versions of API and use the corresponding implementations.

Also, hard-coded HTML documentation is changed to the corresponding URI in the project repo in favor of maintainability.

Related task: https://progress.opensuse.org/issues/62531

OleksandrOrlov committed:
Return JSON containing doc url upon "/" request

The commit replaces hard-coded HTML documentation with the url to the
actual documentation in project repo to keep it in one place and in
up-to-date state without being forced to update in several places.

OleksandrOrlov committed:
Add API version to the resource URIs

rwx788 commented on src/YHttpServer.cc:
```diff
@@ -331,7 +331,10 @@ bool YHttpServer::process_data()
-void YHttpServer::mount(const std::string& path, const std::string &method, YHttpHandler *handler)
+void YHttpServer::mount(std::string path, const std::string &method, YHttpHandler *handler, bool has_api_version)
 {
+    if (has_api_version) {
```
Please, move bracket to the new line.

OleksandrOrlov replied:
Done.

shundhammer commented on src/YHttpServer.h:
```diff
@@ -29,6 +29,8 @@
 #define YUI_AUTH_USER       "YUI_AUTH_USER"
 #define YUI_AUTH_PASSWD     "YUI_AUTH_PASSWD"
 
+#define YUI_API_VERSION     "v1"
```
Better make this something that is easier to compare numerically; something like `01.00` (no `v` prefix). This will make life a lot easier in the future.

lslezak replied:
For the REST API versioning paths like `/api/v1` are very common, I'd stick with that convention.

Numerical comparison does not make much sense here, you cannot know if v3 is (will be) backward compatible with v2 or not. Both server and client API versions needs to match otherwise you are asking for troubles...

shundhammer replied:
Well, okay then. I was thinking about how Qt handles this; they let you go into very deep detail what you can do depending on the exact version.

lslezak commented:
REST API versioning in general is a bit tricky because there is no standard for that and there are several possible solutions/implementations. And sadly each of them has different pros&cons. (See e.g. [here](https://stackoverflow.com/questions/40424720/rest-api-versioning-popular-apis), [here](https://stackoverflow.com/questions/389169/best-practices-for-api-versioning) or [here](https://www.troyhunt.com/your-api-versioning-is-wrong-which-is/).) E.g. the SCC REST API uses the `Accept` header way (see the [documentation](https://scc.suse.com/connect/v4/documentation)).

The question is whether the backward compatibility should be implemented on the server side (serving e.g. both v1 and v2, clients would only talk v1 or v2) or on the client side (the latest server would support only v2 and the clients would support v2 and switch to v1 if the serves used that). I'd prefer the former solution as there might be 3rd party clients out of our control.

The question is how many older versions we should keep working, i.e. should be v3, v2 and v1 be supported or just v3 and v2 or even only the latest one (so you would need to use the client matching the server)?

Do you have some experience with REST API versioning? What would you suggest?

lslezak commented on src/YHttpRootHandler.cc:
```diff
@@ -16,128 +16,28 @@
 #include <boost/algorithm/string/replace.hpp>
+#include <json/json.h>
+
+const std::string YHttpRootHandler::documentation = "https://github.com/libyui/libyui-rest-api/#libyui-embedded-webserver";
```
This would always point to the latest documentation, we would need to create a version specific documentation  `https://github.com/libyui/libyui-rest-api/README.v1.md` or something like that which would be still valid even after we introduce API v2 in the master branch...

OleksandrOrlov replied:
This issue we propose to resolve in the following way: each time when the major version is bumped (i.e. if any backwards incompatible changes are introduced to the API), we create separate branch for the older version and update the `documentation_url` with the appropriate URL. The latest version will be in `master` all the time.

Is this Ok for you?

lslezak commented on src/YHttpRootHandler.cc:
```diff
@@ -16,128 +16,28 @@
-"      </p>"
-"      <hr>"
-"      <h3>Dump Whole Dialog</h3>"
-"      <p>Request: GET <a href='/dialog'>/dialog</a></p>"
```
My idea with the human readable documentation in the document root (`/` path) was that if you do not know how to use the REST API you can simply open the server URL in a browser. And there you could just click that `/dialog` link  there and check the result without any other knowledge about the requested path or format for that.

With the `@port@` variable replacement you could also simply copy&paste the curl command examples to terminal and it would just work.

OleksandrOrlov replied:
@lslezak The idea sounds reasonable, but unfortunately it requires additional efforts on maintenance as it should be duplicated both in README.md and on root (`/` endpoint).

Maybe I just do not know the initial plan on what is the expected use cases for the API. I guess here we should come to consensus on what is the less evil and try to define the target audience:) 

I as a test developer will probably hardly use the links as most of the time I'll send the requests using [Client](https://github.com/jknphy/libyui_client). Newcomers will be able to proceed by `documentation_url` link and also see the required endpoints in case they want to get familiar with the API itself.

Could you please let me know if you are OK with the proposed solution with the link to documentation in JSON, otherwise I'll return the HTML documentation.

lslezak commented on src/YHttpServer.cc:
```diff
@@ -250,7 +250,7 @@ static int onConnect(void *srv, const struct sockaddr *addr, socklen_t addrlen)
 void YHttpServer::start()
 {
-    mount("/", "GET", new YHttpRootHandler());
+    mount("/", "GET", new YHttpRootHandler(), false);
```
I'd put the version info into a separate subpath, e.g. `/version` and keep the root for some human readable HTML documentation. It could simply contain a link to the GitHub documentation to avoid the duplication, but I think it should be something readable for humans...

OleksandrOrlov replied:
Done. I've created the separate endpoint `/version`.

rwx788 commented:
> REST API versioning in general is a bit tricky because there is no standard for that and there are several possible solutions/implementations. And sadly each of them has different pros&cons. (See e.g. [here](https://stackoverflow.com/questions/40424720/rest-api-versioning-popular-apis), [here](https://stackoverflow.com/questions/389169/best-practices-for-api-versioning) or [here](https://www.troyhunt.com/your-api-versioning-is-wrong-which-is/).) E.g. the SCC REST API uses the `Accept` header way (see the [documentation](https://scc.suse.com/connect/v4/documentation)).
> 
> The question is whether the backward compatibility should be implemented on the server side (serving e.g. both v1 and v2, clients would only talk v1 or v2) or on the client side (the latest server would support only v2 and the clients would support v2 and switch to v1 if the serves used that). I'd prefer the former solution as there might be 3rd party clients out of our control.
> 
> The question is how many older versions we should keep working, i.e. should be v3, v2 and v1 be supported or just v3 and v2 or even only the latest one (so you would need to use the client matching the server)?
> 
> Do you have some experience with REST API versioning? What would you suggest?

We have considered all the options, and I guess that we are fine just to support latest version of API on server side. For the older versions of distributions, to execute new tests, we can then request updating libyui stack (which will be included with maintenance updates anyway), or in the worst case - backport some of the new API changes.
Considering limited resources, I guess we will mainly support latest version of the API, and branch client library too.

Approach with headers is a good alternative, in case we add default fallback to the latest API version in case no version is given, so people still could check dialogs with browser or simple curl requests.
I guess we are not that concerned about network overhead here (as need to add API version to all requests), so it's viable option too.

Both of the approaches will allow us support multiple REST API version on both sides, so this design decision is actually independent from versioning itself, I would say.

OleksandrOrlov committed:
Add API version handler

The commit introduces "/version" resource with current API version field
in the response body to allow clients recognize the version.

OleksandrOrlov committed:
Bump version to 0.4.0

lslezak committed:
Improved / handling, added API_v1.md

- For the root page (/) return the HTML or JSON
  document depending on the "Accept" request header
- Merged error code, response body and encoding into a single call

lslezak commented:
I created some improvements in https://github.com/OleksandrOrlov/libyui-rest-api/pull/1 (pull request for pull request).

- For the root page (/) return the HTML or JSON document depending on the "Accept" request header
- Added API v1 documentation in a separate file (`API_v1.md`), when we introduce v2 we will document it also separately so the old link to the v1 documentation will be still valid.

lslezak commented on package/libyui-rest-api.changes:
```diff
@@ -1,3 +1,11 @@
+
+- Replace hard-coded HTML documentation with the url to the actual
+  documentation in project repo
+- Add "/version" endpoint to access to API version
```
Please add a note that the API is versioned now and uses the `/v1/` prefix in the URL path.

rwx788 commented:
@lslezak Hey Ladi! Any chance you can take a look on the PR?

lslezak commented:
Yes, it was waiting for the SP2 branching. I'll later check again...

rwx788 commented:
> Yes, it was waiting for the SP2 branching. I'll later check again...

Alright, thanks!

lslezak reviewed (approved):
LGTM

lslezak commented:
@rwx788 @OleksandrOrlov Thank you!

rwx788 commented:
> @rwx788 @OleksandrOrlov Thank you!

Thanks for your help! Very appreciated! I will resolve conflicts for #5
